opm/grid/GraphOfGridWrappers.cpp

```diff
@@ -208,201 +208,159 @@ void extendGIDtoRank(const GraphOfGrid<Dune::CpGrid>& gog,
+        // iterate once through the original exportList and identify exported wells
+        for (const auto& cellProperties : exportList)
         {
-            int rankToExport = std::get<1>(cellProperties);
-            if (rankToExport!=root)
+            auto pWell = wellMap.find(std::get<0>(cellProperties));
+            if (pWell!=wellMap.end())
             {
-                const auto& [begin, end, wSize] = pWell->second;
-                std::vector<int> wellToExport;
-                wellToExport.reserve(wSize);
-                wellToExport.push_back(*begin);
-                // well ID is its cell of lowest index and is already in the exportList
-                assert(*begin==std::get<0>(cellProperties));
-                for (auto pgID = begin; ++pgID!=end; )
+                int rankToExport = std::get<1>(cellProperties);
+                if (rankToExport!=root) // wells on root are not exported
                 {
-                    // cells in one well have the same attributes (except ID)
-                    std::tuple<int,int,char> wellCell = cellProperties;
-                    std::get<0>(wellCell) = *pgID;
-                    addToList.push_back(wellCell);
-
-                    wellToExport.push_back(*pgID);
+                    const auto& [begin, end, wSize] = pWell->second;
+                    wellsToExport.emplace_back(begin, end, rankToExport);
+                    sizesOfExport[rankToExport] += wSize-1; // one cell is already in the list
+                }
+                wellMap.erase(pWell);
+                if (wellMap.empty())
+                {
+                    break;
                 }
-                exportedWells[rankToExport].push_back(std::move(wellToExport));
-            }
-            wellMap.erase(pWell);
-            if (wellMap.empty())
-            {
-                break;
             }
         }
     }
 
-    // add new cells to the exportList and sort it. It is assumed that exportList starts sorted.
-    std::sort(addToList.begin(), addToList.end());
-    auto origSize = exportList.size();
-    auto totsize = origSize+addToList.size();
-    exportList.reserve(totsize);
-    exportList.insert(exportList.end(), addToList.begin(), addToList.end());
-    std::inplace_merge(exportList.begin(), exportList.begin()+origSize, exportList.end());
+    // create the output: the cells that are missing from importList on non-root ranks
+    // also add new cells to the exportList and sort it
+    exportedCells.resize(cc.size());
+    int addedExportsSize = 0;
+    for (int i=0; i<cc.size(); ++i)
+    {
+        exportedCells[i].reserve(sizesOfExport[i]);
+        addedExportsSize += sizesOfExport[i];
+    }
+    exportList.reserve(exportList.size()+addedExportsSize);
+    for (const auto& pWell : wellsToExport)
+    {
+        const auto& [begin, end, rank] = pWell;
+        // remember to skip the well's first cell which already is in the importList
+        for (auto pgID = begin; ++pgID!=end; )
```

Comment:
Using `begin+1` would be indeed clearer. Unfortunately, it does not compile.

I am storing `begin` as a pointer to well because it gets incremented only once and it felt more natural to do it in the moment we are skipping the first cell.

Reply:
That is surprising. what is the compiler error of
```suggestion
        for (auto pgID = begin + 1;  pgID != end; ++pgID )
```

Reply:
error: no match for ‘operator+’ (operand types are ‘std::tuple_element<0, const std::tuple<std::_Rb_tree_const_iterator<int>, std::_Rb_tree_const_iterator<int>, int> >::type’ {aka ‘std::add_const<std::_Rb_tree_const_iterator<int> >::type’} and ‘int’)
  291 |         for (auto pgID = begin+1; pgID!=end; ++pgID)
      |                          ~~~~~^~
      |                          |     |
      |                          |     int
      |                          std::tuple_element<0, const std::tuple<std::_Rb_tree_const_iterator<int>, std::_Rb_tree_const_iterator<int>, int> >::type {aka std::add_const<std::_Rb_tree_const_iterator<int> >::type}

Reply:
So the problem is storing "const" inside of the tuple. I need to think about this.

Reply:
The iterator in the `set` is not RandomAccess, so arithmetic is not supported. Nevertheless, removing some constness on outer iterator allows us to use
        `for (auto& pgID = ++begin; pgID!=end; ++pgID)`